### PR TITLE
optimize particle shape implementation

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Optimize the implementation of the particle shapes to help the compiler
to create optimized vector code.
Use easy if patterns to give the compiler the freedom to transform the if statements.

# Background

Currently we using something like arithmetic ifs to avoid branch differences during the particle shape evaluation. This will force the compiler to calculate all branches of the if. By refactoring the if conditions we can help the compiler to optimize the code much better. That allows the compiler to use masked assignments, to use branch differences or any other architecture specific optimizations.

thx to INTEL(R) for the [A Guide to
Vectorization with
Intel® C++ Compilers](https://d3f8ykwhia686p.cloudfront.net/1live/intel/CompilerAutovectorizationGuide.pdf)

# Test

- [x] KHI heating
- [x] KHI speed

### CODE REVIEW

**Comment by theZiz** [comment]
Hm, I can't find where in `A Guide to Vectorization with Intel® C++ Compilers` it is stated that your changes make the code better vectorizable. Can you provide me a page number?

**Comment by psychocoderHPC** [comment]
@theZiz section 3.3: `[...]However, “if” statements are allowed if they can be
implemented as masked assignments, which is usually the case. [...] `

File: include/picongpu/particles/shapes/CIC.hpp [hunk 1]
```diff
@@ -58,9 +58,14 @@ struct CIC : public shared_CIC::CIC
...
-            const bool below_1 = (abs_x < float_X(1.0));
+            const bool below_1 = abs_x < float_X(1.0);
+            const float_X onSupport = float_X(1.0) - abs_x;
 
-            return float_X(below_1) * (float_X(1.0) - abs_x);
+            float_X result(0.0);
+            if(below_1)
+                result = onSupport;
+
+            return result;
```
Comment (theZiz) [inline include/picongpu/particles/shapes/CIC.hpp]: I don't get the change for this example as we will have only one return for both branches.

**Comment by theZiz** [comment]
Okay, thanks, I saw this. This explains why to avoid multiple returns, but not why you changed the first code in the commit (where I also made a comment). Did you just not wanted to do the masking "by hand"?

**Comment by psychocoderHPC** [comment]
@slizzered I can't do masking by hand, because this needs asm commands and those are architecture specific. Know the compiler can choose the best fitting asm command or pattern.

**Comment by theZiz** [comment]
Let us keep slizzered out of this. :stuck_out_tongue: 
I got it now. :shipit:

File: include/picongpu/particles/shapes/CIC.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/shapes/CIC.hpp]: offline discussed: I changed this line that this part looks like all other changes

**Comment by psychocoderHPC** [comment]
# Scaling Results

- KHI electrons/positrons
- particle drift in X direction
- global cells:
  - 3D - 192x512x12
  - 2D - 192x512
- SuperCellSize 8x8x4
- 3000 time steps
- 25 ppc/species


## **test case k20:**
- 1 GPU (k20)



| dimension | shape | Esirkepov  dev[s] | Esirkepov  PR[s] | speedup |
| --- | --: | --: | --: | --: |
| 3D | CIC | 787 | 798 | 0.99|
| 3D | TSC | 1944 | 1893 | 1.03 |
| 3D | PCS | 5953 | 5576 | 1.07 |
| 3D | P4S | 19348 | 14322 | 1.35 |
| 2D | CIC |55 | 55 | 1.00 |
| 2D | TSC | 95 | 93 | 1.02 |
| 2D | PCS | 154 | 147 | 1.05 |
| 2D | P4S | 274 | 230 | 1.19 |

| dimension | shape | EmZ  dev[s] | EmZ  PR[s] | speedup |
| --- | --: | --: | --: | --: |
| 3D | CIC | 787 | 714 |1.10 |
| 3D | TSC | 1725 | 1669 | 1.03  |
| 3D | PCS | 5542 | 5215 | 1.06 |
| 3D | P4S | 15144 | 12329 | 1.23 |
| 2D | CIC | 54 | 54 | 1.00 |
| 2D | TSC |94  |93  | 1.01  |
| 2D | PCS | 161 | 145 | 1.11 |
| 2D | P4S |  292| 220 | 1.33 |


## **test case laser:**
- KHI electrons/positrons
- 1 laser node (8MPI ranks)

speedup normalized to dev

| dimension | shape | Esirkepov  dev[s] | Esirkepov  PR[s] | speedup |
| --- | --: | --: | --: | --: |
| 3D | CIC | 3710 | 3480 | 1.07 |
| 3D | TSC | 10865 | 7123 | 1.53 |
| 3D | PCS | 27061*| 17133 | 1.58 |
| 3D | P4S | 53320*  | 52140* | 1.02 |
| 2D | CIC |249 | 237 | 1.05 |
| 2D | TSC | 460 | 349 | 1.32 |
| 2D | PCS | 822 | 510 | 1.61 |
| 2D | P4S | 1224 | 1151 | 1.06 |



**test case k20:**
- KHI electrons/positrons
- 1 laser node (8MPI ranks)


speedup normalized to dev

| dimension | shape | EmZ  dev[s] | EmZ  PR[s] | speedup |
| --- | --: | --: | --: | --: |
| 3D | CIC | 2749 | 2759 | 1.00 |
| 3D | TSC | 7029 | 4199 | 1.67 |
| 3D | PCS | 14633 | 11461 | 1.28 |
| 3D | P4S |  40457*| 40486* | 1.0 |
| 2D | CIC | 249| 216 | 1.15 |
| 2D | TSC | 460 | 295 | 1.56 |
| 2D | PCS | 566 | 371 | 1.53 |
| 2D | P4S | 917 | 874 | 1.05 |


* waltime limit, linear scaled to 3000 steps

**Comment by theZiz** [comment]
Nice!
Over 9000!

**Comment by psychocoderHPC** [comment]
left = numerical heating heating
right = charge conservation

# 3D

## CIC
3d_cic

## TSC
3d_tsc


## PCS
3d_pcs


## P4S
3d_p4s



# 2D

## CIC
2d_cic


## TSC
2d_tsc


## PCS
2d_pcs


## P4S
2d_p4s

File: include/picongpu/particles/shapes/P4S.hpp [hunk 2]
```diff
@@ -103,17 +103,20 @@ struct P4S : public shared_P4S::P4S
...
-            {
-                const bool below_2rd_radius = (abs_x < float_X(0.5));
-                if (below_2rd_radius)
-                    return ff_1st_radius(abs_x);
-                else
-                    return ff_2nd_radius(abs_x);
-            }
-            else
-                return ff_3rd_radius(abs_x);
+            const bool below_2rd_radius = abs_x < float_X(1.5);
```
Comment (ax3l) [inline include/picongpu/particles/shapes/P4S.hpp]: 2nd

File: include/picongpu/particles/shapes/P4S.hpp [hunk 3]
```diff
@@ -103,17 +103,20 @@ struct P4S : public shared_P4S::P4S
...
+            const bool below_1th_radius = abs_x < float_X(0.5);
```
Comment (ax3l) [inline include/picongpu/particles/shapes/P4S.hpp]: 1st

Comment (PrometheusPi) [reply include/picongpu/particles/shapes/P4S.hpp]: and 2nd

**Pull request merged**